8mm evk a53 support

MrVan opened the pull request:
This goal is to support Zephyr runs on top of Jailhouse hypervisor. 

This patchset is the first step to make Zephyr runs on i.MX8MM A53 core from U-Boot.

MrVan committed:
soc: arm64: nxp_imx: Add i.MX8M Mini SoC A53 support

Add minimal SoC support for the NXP i.MX8M Mini series Cortex-A53 Core.

Signed-off-by: Peng Fan <peng.fan@nxp.com>

MrVan committed:
board: arm64: add i.MX8MM EVK A53 board support

Add i.MX8MM EVK A53 board support.
Zephyr is kicked from NS-EL1 U-Boot.

Signed-off-by: Peng Fan <peng.fan@nxp.com>

hakehuang commented on boards/arm/mimx8mm_evk_a53/board.cmake:
```diff
@@ -0,0 +1 @@
+# SPDX-License-Identifier: Apache-2.0
```
need add debugger support
`
board_set_debugger_ifnset(jlink)
board_set_flasher_ifnset(jlink)

board_runner_args(jlink "--device=MIMX8MD6_A53")
include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
`

MrVan replied:
I am not sure, but I not see this "./boards/arm/bcm958402m2_a72/board.cmake" and "boards/arm/qemu_cortex_a53/board.cmake",  Is it really needed?

hakehuang replied:
without that you can not use west flash to download the application. how do you program the board now?

MrVan replied:
Use U-Boot to load/kick the zephyr.bin or use Linux to load it for hypervisor usage. zephyr.bin is stored in sdcard fat partition.

hakehuang replied:
@MrVan  the official way is to use `west flash`

carlocaione replied:
Nope, using U-Boot is fine and `west flash` is totally optional here. Not always you want to use the debugger to load the image also considering that you do not always have a debugger.

hakehuang replied:
then user need extra steps to setup uboot, and as uboot is evovled overtime, this needs some extra documents and efforts.

carlocaione replied:
How to use U-Boot is already documented in this commit and that is enough. Fine to me if we can use `west flash` to flash the board as well but there is no such a thing as "official" way to flash or use the board. @MrVan is this a problem adding support for JLink?

MrVan replied:
I am not familar with JLink. To me, we have sdcard to hold zephyr.bin, zephyr.bin is copied to the sdcard by mounting the sdcard to Linux PC. Could west flash support writing fat partition of a sdcard? Actually we have yocto Linux for image generating.

hakehuang replied:
@MrVan west does not support fat partition write, it is a different way. It may not be officially defined to use west flash, but basically we expect to have an easy to use solution.

carlocaione replied:
> @MrVan west does not support fat partition write, it is a different way. It may not be officially defined to use west flash, but basically we expect to have an easy to use solution. Besides if there no debuger tools supported, how users suppose to debug the program? IMHO, it is not a good enough enablement.

I disagree. U-Boot + SD is indeed an easy solution to use in this case (and pretty standard when dealing with AArch64 boards, much more common than having a J-Link available) and you cannot expect that everybody has access to a debugger / J-Link or that every board has a debug port easily accessible. So, this enablement is good enough for merging.

Now, that said @MrVan, can you add what suggested here https://github.com/zephyrproject-rtos/zephyr/pull/32694#discussion_r583531459 so that `west flash` can be used when a debugger is available? This doesn't prevent the user from using U-Boot and the method you already described.

Please, also update the documentation describing both the methods.

hakehuang reviewed (commented):
suggest to create an issues to explain on how to use this with uboot + jailhouse.

carlescufi commented on boards/arm/mimx8mm_evk_a53/doc/index.rst:
```diff
@@ -0,0 +1,111 @@
+
+i.MX8M Mini LPDDR4 EVK board is based on NXP i.MX8M Mini applications
+processor, composed of a quad Cortex®-A53 cluster and a single Cortex®-M4 core.
+Zephyr OS is ported to run on the Cortex®-M4 core.
```
Did you mean to say that Zephyr already runs on the M4, but now it can also run on the A53? It's a bit confusing as it is written now.

MrVan replied:
It runs on A53 now.

galak commented on soc/arm/nxp_imx/mcimx8mm_a53/Kconfig.defconfig.mimx8mm_a53:
```diff
@@ -0,0 +1,44 @@
+config SOC
+	default "mimx8mm6"
+
+config FLASH_SIZE
```
Would be good for those to be coming in via devicetree and having `zephyr,flash` set.

MrVan replied:
Currently I not have the flash on board, so use fixed value here should be ok? I could update to use your suggested method when we use flash in future.

MrVan replied:
Changed to use device tree.

Mani-Sadhasivam commented:
@MrVan FWIW, I worked on this a while ago but never managed to clean it up and submit the PR:
https://github.com/Mani-Sadhasivam/zephyr/commits/jailhouse

I had to modify the GIC driver to support GIC v2.

carlocaione commented on boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig:
```diff
@@ -0,0 +1,29 @@
+CONFIG_SRAM_BASE_ADDRESS=0x40000000
+CONFIG_XIP=n
+CONFIG_SRAM_SIZE=1024
+CONFIG_ARM64_PA_BITS_48=y
```
you have `CONFIG_ARM64_PA_BITS_36=y` at line 8 up here.

MrVan replied:
Resolved.

carlocaione commented on boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig:
```diff
@@ -0,0 +1,29 @@
+CONFIG_ARM_ARCH_TIMER=y
+
+# Zephyr Kernel Configuration
+CONFIG_SRAM_BASE_ADDRESS=0x40000000
```
This should not be necessary because the value is directly read from the DTS.

carlocaione commented on boards/arm/mimx8mm_evk_a53/mimx8mm_evk_a53_defconfig:
```diff
@@ -0,0 +1,29 @@
+CONFIG_ARM64_PA_BITS_48=y
+
+# Serial Drivers
+CONFIG_SERIAL=y
```
you have `CONFIG_SERIAL=y` but no serial driver enabled?

MrVan replied:
I have this in Kconfig

`
if SERIAL

config UART_MCUX_IUART
        default y

endif # SERIAL
`

carlocaione commented on soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:
```diff
@@ -0,0 +1,49 @@
+	MMU_REGION_FLAT_ENTRY("GIC",
+			      DT_REG_ADDR_BY_IDX(DT_INST(0, arm_gic), 0),
+			      DT_REG_SIZE_BY_IDX(DT_INST(0, arm_gic), 0),
+			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
```
instead of `MT_RW` please use `MT_P_RW_U_RW` or similar.

MrVan replied:
fix in next version

carlocaione commented on soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:
```diff
@@ -0,0 +1,49 @@
+
+	MMU_REGION_FLAT_ENTRY("DRAM",
+			      CONFIG_SRAM_BASE_ADDRESS,
+			      CONFIG_SRAM_SIZE * SZ_1K,
```
Use `KB()` instead of `* SZ_1K`

carlocaione commented on soc/arm/nxp_imx/mcimx8mm_a53/mmu_regions.c:
```diff
@@ -0,0 +1,49 @@
+			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
+
+	MMU_REGION_FLAT_ENTRY("DRAM",
+			      CONFIG_SRAM_BASE_ADDRESS,
```
DRAM vs SRAM. But why do you need this? In general it should not be needed to map the whole SRAM.

MrVan replied:
I am thinking we need memory for IPC usage, between zephyr and Linux.

carlocaione replied:
Ok, but mapping the whole SRAM is probably an overkill. Can you define a memory region to be used for IPC and map that instead?

MrVan replied:
sure. Currently I have not add that piece memory, will add that in future PRs.

MrVan commented:
> @MrVan FWIW, I worked on this a while ago but never managed to clean it up and submit the PR:
> https://github.com/Mani-Sadhasivam/zephyr/commits/jailhouse
> 
> I had to modify the GIC driver to support GIC v2.

Wow. what usecases you are making zephyr on Jailhouse for?

MrVan commented:
updated, please help check.

MrVan commented:
Updated, please help check

carlocaione commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -103,11 +104,31 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
+		break;
+
+	case IMX_CCM_UART2_CLK:
+
```
nit: empty line

carlocaione commented on dts/arm/nxp/nxp_imx8mm_a53.dtsi:
```diff
@@ -0,0 +1,82 @@
+			      IRQ_DEFAULT_PRIORITY>;
+		label = "arch_timer";
+		interrupt-parent = <&gic>;
+	};
```
nit: add empty line here

carlocaione commented on dts/arm/nxp/nxp_imx8mm_a53.dtsi:
```diff
@@ -0,0 +1,82 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
```
given that we have `CONFIG_MP_NUM_CPUS=2` should we have `cpu@1` here as well?

MrVan commented:
@MaureenHelm for the build failure 

> /var/lib/buildkite-agent/zephyr-module-cache/modules/hal/nxp/mcux/devices/MIMX8MM6/MIMX8MM6_cm4.h:235:10: fatal error: core_cm4.h: No such file or directory
--
  | 235 \| #include "core_cm4.h"                  /* Core Peripheral Access Layer */
  | \|          ^~~~~~~~~~~~

Should I send a PR to nxp hal?

MaureenHelm reviewed (approved):
> @MaureenHelm for the build failure
> 
> > /var/lib/buildkite-agent/zephyr-module-cache/modules/hal/nxp/mcux/devices/MIMX8MM6/MIMX8MM6_cm4.h:235:10: fatal error: core_cm4.h: No such file or directory
> > --
> >   | 235 | #include "core_cm4.h"                  /* Core Peripheral Access Layer */
> >   | |          ^~~~~~~~~~~~
> 
> Should I send a PR to nxp hal?

Yes. Please also send a PR to https://github.com/NXPmicro/mcux-sdk

MrVan commented:
NXP HAL PR: https://github.com/zephyrproject-rtos/hal_nxp/pull/80

Add a few more commits to address build warning for ARM64 and support Jailhouse hypervisor.

MrVan commented:
https://github.com/NXPmicro/mcux-sdk/pull/11 created.

MrVan committed:
dts: arm64: nxp: add i.MX8MM A53 device tree

Add i.MX8MM A53 device tree

Signed-off-by: Peng Fan <peng.fan@nxp.com>

MrVan commented:
Finally find some time to rework to drop NXP HAL MCU dependency, because this is for Cortex-A usage, and NXP not has Cortex-A HAL SDK support, so added a new uart driver for i.MX8MM A53 usage without depends on NXP HAL MCU SDK. Hope this could make the PR in, thanks.

MrVan committed:
CODEOWNERS: update CODEOWNERS for i.MX8MM

Update CODEOWNERS for i.MX8MM

Signed-off-by: Peng Fan <peng.fan@nxp.com>

MrVan commented:
@carlocaione @ioannisg @MaureenHelm @carlescufi I think the build failure are common for ARM64? So this PR is ready to be merged? Thanks.

carlocaione commented on boards/arm64/mimx8mm_evk_a53/mimx8mm_evk_a53.yaml:
```diff
@@ -0,0 +1,12 @@
+identifier: mimx8mm_evk_a53
+name: NXP i.MX8M Mini EVK A53
+type: mcu
+arch: arm
```
The failure you see on CI are caused by this. It should be `arm64`.

MrVan replied:
oh. I missed to change this after aarch64 has a seperate folder.

carlocaione reviewed (changes requested):
Let's wait for the CI to be all green.

MrVan commented:
@carlocaione CI passed.

MrVan commented:
@carlocaione @ioannisg @MaureenHelm @carlescufi Could this PR get merged now?

carlocaione commented:
@MrVan we are in code freeze. This will be merged after release.

MrVan commented:
@carlocaione ok, thanks for the info.

galak reviewed (dismissed):
@MaureenHelm we need to figure out how to handle the drivers for this.  I don't think it makes sense to have duplicated HAL & non-HAL drivers.  Doubles our support efforts.

MrVan commented:
> @MaureenHelm we need to figure out how to handle the drivers for this. I don't think it makes sense to have duplicated HAL & non-HAL drivers. Doubles our support efforts.

The effort to reuse NXP SDK https://github.com/NXPmicro/mcux-sdk/pull/11 has been rejected.

carlocaione commented:
@MrVan what do you plan to do with this?

MrVan commented:
> @MrVan what do you plan to do with this?

After https://github.com/zephyrproject-rtos/hal_nxp/pull/80 ready, I'll rework this PR.

arvid-r commented:
Let me know if you need any testing related to this. I have the relevant boards (I believe) and would really like to see Jailhouse/Zephyr on arm64 at some point, so I'll try to help if I can.

dleach02 commented:
@MrVan, the NXP HAL switch PR has been merged. When you get the chance please rebase and verify.

MrVan commented:
> @MrVan, the NXP HAL switch PR has been merged. When you get the chance please rebase and verify.

@dleach02 could your please share me the merged PR link?

MrVan commented:
> > @MrVan, the NXP HAL switch PR has been merged. When you get the chance please rebase and verify.
> 
> @dleach02 could your please share me the merged PR link?
@dleach02
BTW: you mean i.MX8MM A53 port has been supported in NXP HAL?

MrVan commented:
> > @MrVan, the NXP HAL switch PR has been merged. When you get the chance please rebase and verify.
> 
> @dleach02 could your please share me the merged PR link?
@dleach02
BTW: you mean i.MX8MM A53 port has been supported in NXP HAL?

MrVan commented:
Update the PR, ram console is used here to avoid the dependency on nxp hal which block the PR for long time. After NXP HAL finally support ARM64, we could switch back. I am not sure why there is a twister build failure, it is caused this PR?

dleach02 commented:
@MrVan, Twister is showing failure in imx8mm_evk_jailhouse with samples/subsys/shell/shell_module/sample.shell.shell_module

MrVan committed:
board: arm64: imx8mm_evk: add a dummy file

Because we currently use RAM_CONSOLE for now, and there is actually
no device in the build, so there is twister build failure,
add a dummy file to avoid build failure.

Signed-off-by: Peng Fan <peng.fan@nxp.com>

MrVan commented:
all check passed, please help check whether you are ok with this PR now. Thanks.

dleach02 commented:
@galak, can you revisit your change request to determine if the author has addressed your request?

MrVan commented:
The major concern from @galak was the PR(previous version) has its own uart driver, not reusing NXP HAL. This conern has been addressed for quite some time, so any more comments from project owners?

Mani-Sadhasivam commented on dts/arm64/nxp/nxp_imx8mm_a53.dtsi:
```diff
@@ -0,0 +1,98 @@
+		interrupts = <GIC_SPI 27 IRQ_TYPE_LEVEL 0>;
+		interrupt-names = "irq_0";
+		interrupt-parent = <&gic>;
+		clocks = <&ccm IMX_CCM_UART_CLK 0x6c 24>;
```
The register offsets are wrong. Even in the M4 dts I did the same thing. Since the offsets were not used by the clock driver, these are irrelevant but we should fix the CCM binding to use `#clock-cells = <1>`.

This will also require changing a few dts, but that shouldn't be a big thing.

MrVan replied:
If you think ccm binding need to change to `#clock-cells = <1>`, I could create the other PR to address it. But in this PR, I would not. I am not sure what 0x6c/24 means here, I just refer the m4 dtsi.

Mani-Sadhasivam replied:
Okay. Let's handle it later.

Mani-Sadhasivam commented on soc/arm64/nxp_imx/mcimx8mm_a53/soc_a53.c:
```diff
@@ -16,7 +16,7 @@
  * @return 0
  */
-static int mimx8mm_a53_init(struct device *arg)
+static int mimx8mm_a53_init(const struct device *arg)
```
This change should be part of SoC commit.

MrVan replied:
ok, I could squash the cleanup commit.

Mani-Sadhasivam replied:
please do so!

dleach02 commented:
@galak,

> @MaureenHelm we need to figure out how to handle the drivers for this. I don't think it makes sense to have duplicated HAL & non-HAL drivers. Doubles our support efforts.

Is this comment/block still valid?

MrVan commented:
> @galak,
> 
> > @MaureenHelm we need to figure out how to handle the drivers for this. I don't think it makes sense to have duplicated HAL & non-HAL drivers. Doubles our support efforts.
> 
> Is this comment/block still valid?

Since no response. I am not sure people still happy to review this one year pending PR. I would create a new PR to restart the work.

carlocaione commented:
@MrVan I don't think that's needed. Can you rebase? I'll try to push this.

gmarull commented on soc/arm64/nxp_imx/imx8m/soc.h:
```diff
@@ -0,0 +1,19 @@
+#include <sys/util.h>
+#include <toolchain.h>
+
+#ifndef _ASMLANGUAGE
+#include <devicetree.h>
+#endif
```
this is not required, ref https://github.com/zephyrproject-rtos/zephyr/pull/41544

gmarull commented on soc/arm64/nxp_imx/imx8m/soc_a53.c:
```diff
@@ -0,0 +1,27 @@
+#include <init.h>
+#include <arch/cpu.h>
+
+/**
+ * @brief Perform basic hardware initialization at boot.
+ *
+ * This needs to be run from the very beginning.
+ * So the init priority has to be 0 (zero).
+ *
+ * @return 0
+ */
+static int mimx8mm_a53_init(const struct device *arg)
+{
+
+	/* TODO */
+
+	return 0;
+}
+
+SYS_INIT(mimx8mm_a53_init, PRE_KERNEL_1, 0);
```
can this be added when this does something?

MrVan replied:
Dropped.

gmarull commented on boards/arm64/imx8mm_evk/dummy.c:
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2021 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef CONFIG_UART_MCUX_IUART
+
+#define DT_DRV_COMPAT nxp_imx_iuart
+
+#include <device.h>
+#include <drivers/uart.h>
+#include <drivers/clock_control.h>
+#include <errno.h>
+#include <soc.h>
+
+static int dummy_init(const struct device *dev)
+{
+	return 0;
+}
+
+#define IUART_MCUX_INIT(n)						\
+									\
+	DEVICE_DT_INST_DEFINE(n,					\
+			    &dummy_init,				\
+			    NULL,					\
+			    NULL,					\
+			    NULL,					\
+			    PRE_KERNEL_1,				\
+			    CONFIG_SERIAL_INIT_PRIORITY,		\
+			    NULL);
+
+DT_INST_FOREACH_STATUS_OKAY(IUART_MCUX_INIT)
+
+#endif
```
what's going on here?

MrVan replied:
This file is not needed after rebase, dropped.

MrVan replied:
Seems CI not happy if this file dropped. Because UART is not enabled for now, so CI report.
samples/subsys/shell/shell_module/src/uart_reinit.c:19: undefined reference to `__device_dts_ord_9'

Bring it back

gmarull reviewed (changes requested):
A few questions

MrVan committed:
board: arm64: imx8mm_evk: add jailhouse hypervisor inmate support

This is to let zephyr run in an inmate cell on top of Jailhouse
hypervisor.

The step as below:
jailhouse enable imx8mm.cell
ailhouse cell create imx8mm-zephyr.cell
jailhouse cell load 1 zephyr.bin -a 0x93c00000
jailhouse cell start 1

Currently we have not use loader to kick zephyr, later we will
use jailhouse loader to kick zephyr.

Signed-off-by: Peng Fan <peng.fan@nxp.com>

MrVan commented:
@carlocaione @MaureenHelm  sorry for late reply, just back from vocation. I have addressed all the comments.

gmarull commented on soc/arm64/nxp_imx/imx8m/soc.h:
```diff
@@ -0,0 +1,21 @@
+#ifndef SOC_H
+#define SOC_H
+
+#include <sys/util.h>
+#include <toolchain.h>
+
+#ifndef _ASMLANGUAGE
+
+#include <device.h>
+
+#endif /* !_ASMLANGUAGE */
+
+/* TODO: Place Holder */
```
none seem to be required, please, justify them. Maybe some other files are not including necessary headers?

carlocaione commented on boards/arm64/imx8mm_evk/imx8mm_evk_jailhouse.yaml:
```diff
@@ -0,0 +1,12 @@
+identifier: imx8mm_evk_jailhouse
+name: NXP i.MX8M Mini EVK A53 Jailhouse Inmate Cell
```
can you add the info that is now in the commit message in the README?

MrVan replied:
Add info to doc/index.rst.

gmarull commented on soc/arm64/nxp_imx/imx8m/soc.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef SOC_H
+#define SOC_H
+
+#include <sys/util.h>
+#include <toolchain.h>
+
+/* TODO: Place Holder */
```
Why are these required? Likely some other files not including what they need?

MrVan replied:
I just follow other arm64 platforms. I think soc.h is a central place for soc related header files, currently we have not add specific device support, so leave todo for now.

gmarull replied:
> I just follow other arm64 platforms. I think soc.h is a central place for soc related header files, currently we have not add specific device support, so leave todo for now.

That doesn't mean others do it right, really. Ref https://github.com/zephyrproject-rtos/zephyr/issues/41543

MrVan replied:
After a check, "arch/arm64/core/smp.c" includes soc.h, so could we refactor the header file later?

gmarull replied:
Yes, fair enough. Could be part of #41543 work.

MrVan commented:
@carlocaione @MaureenHelm Ready for merge now?

carlocaione commented:
> @carlocaione @MaureenHelm Ready for merge now?

Yes, but we are in stabilization period now. This will be merged when 3.0 will be released.

mbolivar-nordic commented on soc/arm64/nxp_imx/imx8m/mmu_regions.c:
```diff
@@ -0,0 +1,49 @@
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_RW | MT_NS),
+
+	MMU_REGION_FLAT_ENTRY("ANA_PLL",
+			      DT_REG_ADDR(DT_INST(0, nxp_imx_ana)),
+			      DT_REG_SIZE(DT_INST(0, nxp_imx_ana)),
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_RW | MT_NS),
+
+	MMU_REGION_FLAT_ENTRY("UART",
+			      DT_REG_ADDR(DT_INST(1, nxp_imx_iuart)),
+			      DT_REG_SIZE(DT_INST(1, nxp_imx_iuart)),
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_RW | MT_NS),
```
Hard-coded instance numbers like this are incorrect to use. Please use an alternate mechanism to specify exactly the nodes you want. Read the DT_INST docstring for more details.

MrVan replied:
Could we use a follow up patch to address this? I just follow qemu platform here. @carlocaione  do you have any suggestions?

Mani-Sadhasivam replied:
Considering the amount of time this PR has been in an open state (1 year), I think we should merge this and follow up with PRs for fixing issues like this. @mbolivar-nordic what do you think?

mbolivar-nordic replied:
No, sorry, I don't want to see incorrect usage of the instance API merged.

MrVan replied:
@mbolivar-nordic I am not experts in zephyr, so any suggestions what I should use?

MrVan commented:
A bit werid, I have updated my 8mm-v1 branch, but this PR not update automatically.

MrVan commented:
> A bit werid, I have updated my 8mm-v1 branch, but this PR not update automatically.

I am not sure why PR not update automatically. what could I do here? Close this PR and reopen a new one?

stephanosio commented:
> > A bit werid, I have updated my 8mm-v1 branch, but this PR not update automatically.
> 
> I am not sure why PR not update automatically. what could I do here? Close this PR and reopen a new one?

@MrVan It looks like something broke on the GitHub side -- we have seen the same issue in other PRs lately. Please open a new PR.
